File: PCA/amplapi/doc/dotnet/_downloads/TrackingModel.cs
using ampl;
using ampl.Entities;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Examples
{
  public class TrackingModel
  {
    /// <summary>
    /// This example shows a simple AMPL iterative procedure implemented in AMPL
    /// </summary>
    /// <param name="args">
    /// </param>
    public static int Main(string[] args)
    {
      string modelDirectory = ((args != null) && (args.Length > 0)) ? args[0]
            : "../../models";
      string solver = ((args != null) && (args.Length > 1)) ? args[1] : null;

      /*
      // If the AMPL installation directory is not in the system search path:
      ampl.Environment env = new ampl.Environment(
        "full path to the AMPL installation directory");
      // Create an AMPL instance
      using (AMPL a = new AMPL(env)) {}
      */

      // Create an AMPL instance
      using (AMPL ampl = new AMPL())
      {
        if (solver != null) ampl.SetOption("solver", solver);
        modelDirectory += "/tracking";

        // Load the AMPL model from file
        ampl.Read(modelDirectory + "/tracking.mod");
        // Read data
        ampl.ReadData(modelDirectory + "/tracking.dat");
        // Read table declarations
        ampl.Read(modelDirectory + "/trackingbit.run");
        // set tables directory (parameter used in the script above)
        ampl.GetParameter("data_dir").Set(modelDirectory);
        // Read tables
        ampl.ReadTable("assets");
        ampl.ReadTable("indret");
        ampl.ReadTable("returns");

        var hold = ampl.GetVariable("hold");
        Parameter ifinuniverse = ampl.GetParameter("ifinuniverse");

        // Relax the integrality
        ampl.SetOption("relax_integrality", true);
        // Solve the problem
        ampl.Solve();
        Console.Write("QP objective value {0}\n",
          ampl.GetObjectives().First().Value);

        double lowcutoff = 0.04;
        double highcutoff = 0.1;

        // Get the variable representing the (relaxed) solution vector
        DataFrame holdvalues = hold.GetValues();
        List<double> toHold = new List<double>(holdvalues.NumRows);
        // For each asset, if it was held by more than the highcutoff, forces it in the model, if
        // less than lowcutoff, forces it out
        foreach (var value in holdvalues.GetColumn("hold"))
        {
          if (value.Dbl < lowcutoff)
            toHold.Add(0);
          else if (value.Dbl > highcutoff)
            toHold.Add(2);
          else
            toHold.Add(1);
        }
        // uses those values for the parameter ifinuniverse, which controls which stock is included
        // or not in the solution
        ifinuniverse.SetValues(toHold.ToArray());

        // Get back to the integer problem
        ampl.SetOption("relax_integrality", false);
        // Solve the (integer) problem
        ampl.Solve();
        Console.Write("QMIP objective value {0}\n",
        ampl.GetObjectives().First().Value);
      }
      return 0;
    }
  }
}

File: PCA/amplapi/doc/cpp/_downloads/trackingmodel.cc
#include "ampl/ampl.h"

#include <vector>
#include <iostream>

int main(int argc, char **argv) {
  try {
    // Create an AMPL instance
    ampl::AMPL ampl;

    /*
    // If the AMPL installation directory is not in the system search path:
    ampl::Environment env("full path to the AMPL installation directory");
    ampl::AMPL ampl(env);
    */

    if (argc > 1)
      ampl.setOption("solver", argv[1]);

    std::string modelDirectory(argc == 3 ? argv[2] : "../models");
    modelDirectory += "/tracking";

    // Load the AMPL model from file
    ampl.read(modelDirectory + "/tracking.mod");
    // Read data
    ampl.readData(modelDirectory + "/tracking.dat");
    // Read table declarations
    ampl.read(modelDirectory + "/trackingbit.run");
    // Set tables directory (parameter used in the script above)
    ampl.getParameter("data_dir").set(modelDirectory);
    // Read tables
    ampl.readTable("assets");
    ampl.readTable("indret");
    ampl.readTable("returns");

    ampl::Variable hold = ampl.getVariable("hold");
    ampl::Parameter ifinuniverse = ampl.getParameter("ifinuniverse");

    // Relax the integrality
    ampl.setBoolOption("relax_integrality", true);
    // Solve the problem
    ampl.solve();
    std::cout << "QP objective value " << ampl.getObjectives().begin()->value() << "\n";

    double lowcutoff = 0.04;
    double highcutoff = 0.1;

    // Get the variable representing the (relaxed) solution vector
    ampl::DataFrame holdvalues = hold.getValues();
    std::vector<double> toHold;
    toHold.reserve(holdvalues.getNumRows());
    // For each asset, if it was held by more than the highcutoff,
    // forces it in the model, if less than lowcutoff, forces it out
    for (auto value : holdvalues.getColumn("hold"))
    {
      if (value.dbl() < lowcutoff)
        toHold.push_back(0);
      else if (value.dbl() > highcutoff)
        toHold.push_back(2);
      else
        toHold.push_back(1);
    }

    // uses those values for the parameter ifinuniverse, which controls
    // which
    // stock is included or not in the solution
    ifinuniverse.setValues(&toHold[0], toHold.size());

    // Get back to the integer problem
    ampl.setBoolOption("relax_integrality", false);
    // Solve the (integer) problem
    ampl.solve();
    std::cout << "QMIP objective value " <<
    ampl.getObjectives().begin()->value() << "\n";
    return 0;
  } catch (const std::exception &e) {
    std::cout << e.what() << "\n";
    return 1;
  }
}
